### ISSUE TITLE
Unusually high GPU memory usage when using `accelerate launch`

### ISSUE THREAD

**Issue by RaymondLi0** [huggingface/accelerate#1336]
### System Info

```Shell
- `Accelerate` version: 0.18.0
- Platform: Linux-5.4.0-139-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Numpy version: 1.22.2
- PyTorch version (GPU?): 1.13.0a0+936e930 (True)
- `Accelerate` default config:
	- compute_environment: LOCAL_MACHINE
	- distributed_type: NO
	- mixed_precision: bf16
	- use_cpu: False
	- num_processes: 1
	- machine_rank: 0
	- num_machines: 1
	- gpu_ids: all
	- rdzv_backend: static
	- same_network: True
	- main_training_function: main
	- downcast_bf16: no
	- tpu_use_cluster: False
	- tpu_use_sudo: False
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [X] My own task or dataset (give details below)

### Reproduction

I'm not very familiar with `accelerate`, but run into some issues trying to load a 15B-parameter model in bf16.
Running the following snippet `debug_accelerate.py` with an A-100 (80GB) GPU:
```python
import subprocess
from accelerate import Accelerator
import torch
from transformers import AutoModelForCausalLM


def gpu_utilization():
    return subprocess.run(["nvidia-smi"], capture_output=True).stdout.decode('utf-8')


print(gpu_utilization())
accelerator = Accelerator()

model = AutoModelForCausalLM.from_pretrained(
    "bigcode/large-model",
    # "gpt2",
    use_auth_token=True,
)
print("After loading model")
print(gpu_utilization())

print("Accelerate prepare...")
model = accelerator.prepare(model)
print(gpu_utilization())

print("Generate ...")
input_ids = torch.ones((1, 10), dtype=torch.long).cuda()
generated_tokens = accelerator.unwrap_model(model).generate(
                input_ids=input_ids,
                num_return_sequences=1,
            )
print(gpu_utilization())
```

There are several issues:

1. When running `python debug_accelerate.py`, it uses 60GB of GPU memory instead of the expected 30GB (15B-parameter model in bf16)
2. `python debug_accelerate.py` runs till the end. But `accelerate launch debug_accelerate.py` raises a cuda-OOM at `accelerator.unwrap_model(model).generate`
3. If I run `accelerate launch --multi_gpu scripts/debug_accelerate.py` instead, it crashes even before, at `accelerator.prepare(model)` with a cuda OOM

### Expected behavior

1. This script should not use more than 30GB of GPU-memory since it loads 15B parameters in bf16 (running `model.bfloat16().cuda()` in another script uses 30GB of GPU-mem)
2. I expected to see the same memory usage whether I use `accelerate launch debug_accelerate.py` or `python debug_accelerate.py`
3. I did not expect `--multi_gpu` to increase the memory usage when preparing the model.

**Comment by sgugger** [comment]
Where do you set your model to bfloat16?

**Comment by RaymondLi0** [comment]
The accelerate config has `- mixed_precision: bf16`, so `accelerator.prepare(model)` should set it to bfloat16 IIUC?

**Comment by sgugger** [comment]
Nope, this is mixed precision training. So, there is a copy of the model in float32 and one in bfloat16. Full training in bfloat16 is not stable enough to converge.

**Comment by RaymondLi0** [comment]
Oh I see, makes sense, thanks!
Then is it expected that in that setting I'm able to `prepare` the model, but `generate` runs OOM (point 2)?

**Comment by sgugger** [comment]
The accumulation of both training and evaluation may tip you over the memory available. But if you can try on that setup, you can definitely do inference (unless you are putting a craze max length) on its own.

**Comment by loubnabnl** [comment]
I have a related question, I'm using accelerate just for inference with the same model (no training) and running this code
```python
accelerator = Accelerator()

model = AutoModelForCausalLM.from_pretrained(
    "bigcode/large-model",
    use_auth_token=True,
)
model = model.bfloat16()
print(f"Before accelerate prepare GPU 0: {gpu_utilization(0)}")
print(f"Before accelerate prepare GPU 1: {gpu_utilization(1)}")

model = accelerator.prepare(model)

print(f"After accelerate prepare on GPU 0: {gpu_utilization(0)}")
print(f"After accelerate prepare on GPU 1: {gpu_utilization(1)}")
```
When configured without distributed training to run on 1GPU the memory usage is consistent with 31GB on GPU 0.
```
Before accelerate prepare GPU 0: GPU Memory Total: 671.875 MB    
After accelerate prepare on GPU 0: GPU Memory Total: 31290.875 MB
```

However when in a Multi-GPU mode, the same code takes 61GB on each GPU. Shouldn't `accelerate.prepare` just copy the 31GB model to both GPUs? Or is the bf16 not set properly?
```
Before accelerate prepare GPU 0: GPU Memory Total: 671.875 MB
Before accelerate prepare GPU 1: GPU Memory Total: 671.875 MB
Before accelerate prepare GPU 0: GPU Memory Total: 1982.875 MB
Before accelerate prepare GPU 1: GPU Memory Total: 671.875 MB
After accelerate prepare on GPU 0: GPU Memory Total: 61186.875 MB
After accelerate prepare on GPU 0: GPU Memory Total: 61186.875 MB
After accelerate prepare on GPU 1: GPU Memory Total: 61186.875 MB
After accelerate prepare on GPU 1: GPU Memory Total: 61186.875 MB
```

This is the setup (same for no distributed):
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: no
        - use_cpu: False
        - num_processes: 2
        - machine_rank: 0
        - num_machines: 1
        - gpu_ids: 0,1
        - rdzv_backend: static
        - same_network: True
        - main_training_function: main
        - downcast_bf16: no
        - tpu_use_cluster: False
        - tpu_use_sudo: False
        - tpu_env: []

**Comment by sgugger** [comment]
This is because when you send your model to prepare, it is wrapped in the `DistributedDataParallel` class which will reserve memory for the gradients it will need to reduce-average at the end of the backward pass. The only way around this is to not send your model to `accelerator.prepare` when only doing eval, as you shouldn't need it the wrapper.

**Comment by loubnabnl** [comment]
Ah ok I see thanks!
I actually need to send the model to `accelerator.prepare` to feed a dataloader and do distributed text generation for inference as it's done [here](https://github.com/huggingface/transformers/blob/f143037789288ba532dada934a118e648e715738/examples/research_projects/codeparrot/scripts/human_eval.py#L194) (I didn't include it in my example to just reproduce the behavior). In this case I will always need at least 61GB and have bf16 turned on?

**Comment by sgugger** [comment]
You should be able to only send your dataloader to the `prepare` method in this example and still perform distributed inference. The model only needs to be wrapped in `DistributedDataParallel` for training.

**Comment by loubnabnl** [comment]
That worked, thanks!
